Push Notification Changes + Project Fixes + Dark Mode Support

droobah opened the pull request:
- Adds support for Dark Mode in iOS 13+
- Fixes absolute paths in project
- Includes changes for push notifications for the following
  1. Push notifications for all app states (foreground, background, terminated)
  2. Push notification actions (requires change to FirebaseProvider.py to set apns->payload->aps->category = PUSH_AUTHENTICATION in the message)
  3. Push notification actions work from watchOS as well

Commit:
- fixed UI issue with dismiss button displaying
- changed colors to work with dark mode on iOS 13
- fixed file relative path issues to allow compiling

Commit:
- updated toast color to work in dark mode

Commit:
Presenter.swift
- restructured Presenter to be a singleton
AppDelegate.swift
- added functionality to AppDelegate to handle push notifications if app is in a terminated state
Protocols.swift
- updated protocols to match restructure
NotificationManager.swift
- removed unneeded GCD for push notification registration
- added 'Allow' action to incoming push notifications (requires change to FirebaseProvider.py)
- handles notification center actions in all app states
PresenterEndpointCallback.swift
- remove pending auths upon failure  to avoid infinite loop
TableViewController.swift
- updated to use Presenter singleton
- crash when handling notification actions in background fixed
Utilities.swift
- changed to os_log for logging
Podfile
- updated to follow Google's recommendations and latest versions

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
- Readded TokenOld.swift to project file

Commit:
- Moved TokenOld.swift file
- Updated project version to 2.0.1

Commit:
- added possible workaround to random token deletion that occurs

nilsbehlen commented on privacyIDEA Authenticator.xcodeproj/project.pbxproj:
```diff
@@ -825,16 +809,16 @@
 					"@executable_path/Frameworks",
 				);
-				PRODUCT_BUNDLE_IDENTIFIER = privacyidea.authenticator;
+				PRODUCT_BUNDLE_IDENTIFIER = net.cowism.privacyidea.authenticator;
```
we probably don't want a changed identifier

droobah replied:
Yeah, sorry, I had to change that to get it to compile and forgot to revert/cherry-pick the change.

nilsbehlen reviewed (commented):
can u add a PR to privacyidea aswell for changing the message payload?

nilsbehlen commented:
I tried it out and it looks good to me, but i didn't get notifications in terminated state (ios 12.4). Are the changes to the firebaseprovider.py needed for that?

droobah commented:
> I tried it out and it looks good to me, but i didn't get notifications in terminated state (ios 12.4). Are the changes to the firebaseprovider.py needed for that?

Yes, the firebaseprovider.py changes are needed for notifications to work in a terminated state.  The changes switch from silent push to alert notifications with all of the required/recommended APNS flags.  I need to do a couple tests to ensure it doesn't affect the Android push notifications.  Once I confirm it won't affect the Android app, I'll toss up a PR for that.  In the meantime, I put the payload change in this comment so you can test yourself.

Also, there appears to be a timing bug that pops up where the loadPIPublicKey function will fail between 10-25% of the time when the app is in the background/terminated during a response to a push notification action.  I'm currently working on that and hopefully will update this PR today with the fix.

```
        fcm_message = {
            "message": {
                        "data": data,
                        "token": firebase_token,
                        "notification": {
                                         "title": data.get("title"),
                                         "body": data.get("question"),
                                        },
                         "android": {
                                     "priority": "HIGH",
                                     "ttl": "120s",
                                     "notification": {
                                                      "sound": "default",
                                                      "tag": "privacyidea.pushtoken",
                                                      "notification_priority": "PRIORITY_MAX"
                                                     },
                                     "fcm_options": { "analytics_label": "AndroidPushToken" }
                                    },
                         "apns": {
                                  "headers": {
                                              "apns-priority": "10",
                                              "apns-push-type": "alert",
                                              "apns-collapse-id": "privacyidea.pushtoken",
                                              "apns-expiration": str(int(time.time())+120)
                                             },
                                  "payload": {
                                              "aps": {
                                                      "sound": "default",
                                                      "category": "PUSH_AUTHENTICATION"
                                                     }
                                             },
                                  "fcm_options": { "analytics_label": "iOSPushToken" }
                                 }
                       }
            }

```

Commit:
Storage.swift
- changed accessibility to .accessibleAfterFirstUnlock to resolve the issue of approving background notifications via watch not working properly
(all push tokens will need to be re-enrolled to fix the issue)

droobah commented:
Okay, the background bug is fixed.  Had to change the KeychainSwift accessibility to allow background access.  Any existing push tokens will need to be re-enrolled to fix the issue.

Commit:
- fixed misspelling

Commit:
- resolved random token deletion bug caused by attempting to access protected data in the background

Commit:
- removed unneeded protected data check loop

droobah committed:
Merge branch 'master' into master

droobah commented:
Okay, I believe I'm done making changes.  Everything appears to be working swimmingly.

nilsbehlen commented:
thanks alot for your help!
